packages/blockfrost/src/blockfrostChainHistoryProvider.ts

```diff
@@ -0,0 +1,305 @@
+    return orderBy(
+      [
+        ...(pool_retire_count ? await fetchPoolRetireCerts(hash) : []),
+        ...(pool_update_count ? await fetchPoolUpdateCerts(hash) : []),
+        ...(mir_cert_count ? await fetchMirCerts(hash) : []),
+        ...(stake_cert_count ? await fetchStakeCerts(hash) : []),
+        ...(delegation_count ? await fetchDelegationCerts(hash) : [])
+      ],
+      (cert) => cert.cert_index
+    ).map((cert) => omit(cert, 'cert_index') as Cardano.Certificate);
+  };
+
+  const fetchJsonMetadata = async (txHash: Cardano.TransactionId): Promise<Cardano.TxMetadata | null> => {
+    try {
+      const response = await blockfrost.txsMetadata(txHash.toString());
+      // Not sure if types are correct, missing 'label', but it's present in docs
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      return blockfrostMetadataToTxMetadata(response as any);
+    } catch (error) {
+      if (formatBlockfrostError(error).status_code === 404) {
+        return null;
+      }
+      throw error;
+    }
+  };
+
+  // eslint-disable-next-line unicorn/consistent-function-scoping
+  const parseValidityInterval = (num: string | null) => Number.parseInt(num || '') || undefined;
+
+  const fetchTransaction = async (
+    hash: Cardano.TransactionId,
+    protocolParameters: ProtocolParametersRequiredByWallet
+  ): Promise<Cardano.TxAlonzo> => {
+    const { inputs, outputs, collaterals } = BlockfrostToCore.transactionUtxos(
+      await blockfrost.txsUtxos(hash.toString())
+    );
+    const response = await blockfrost.txs(hash.toString());
+    const metadata = await fetchJsonMetadata(hash);
+    const certificates = await fetchCertificates(response);
+    const withdrawals = await fetchWithdrawals(response);
+    return {
+      auxiliaryData: metadata
+        ? {
+            body: { blob: metadata }
+          }
+        : undefined,
+      blockHeader: {
+        blockNo: response.block_height,
+        hash: Cardano.BlockId(response.block),
+        slot: response.slot
+      },
+      body: {
+        certificates,
+        collaterals,
+        fee: BigInt(response.fees),
+        inputs,
+        mint: await fetchMint(response),
+        outputs,
+        validityInterval: {
+          invalidBefore: parseValidityInterval(response.invalid_before),
+          invalidHereafter: parseValidityInterval(response.invalid_hereafter)
+        },
+        withdrawals
+      },
+      id: hash,
+      implicitCoin: Cardano.util.computeImplicitCoin(protocolParameters, { certificates, withdrawals }),
+      index: response.index,
+      txSize: response.size,
+      witness: {
+        redeemers: await fetchRedeemers(response),
+        signatures: new Map() // not available in blockfrost
+      }
+    };
+  };
+
+  const currentWalletProtocolParameters = async (): Promise<ProtocolParametersRequiredByWallet> => {
```

Comment:
This one is also in `blockfrostWalletProvider` but we need it here too. This one is very simple but I'm thinking that in the future we could have more functions like this needed by several providers, maybe we could have some sort of shared base provider where you can do something like this instead:

```ts
export const blockfrostChainHistoryProvider = (
  blockfrost: BlockFrostAPI,
  logger = dummyLogger
): ChainHistoryProvider => {
  const { currentWalletProtocolParameters } = baseBlockfrostProvider(blockfrost);

  const transactionsByHashes: ChainHistoryProvider['transactionsByHashes'] = async (hashes) => {
    const protocolParameters = await currentWalletProtocolParameters();
    return Promise.all(hashes.map((hash) => fetchTransaction(hash, protocolParameters)));
  }

  return { transactionsByHashes };
}
```

Does it make sense?

Reply:
If we moved `currentWalletProtocolParameters` to `NetworkInfoProvider`, then it could be:

``` ts
export const blockfrostChainHistoryProvider = (
  blockfrost: BlockFrostAPI,
  networkInfoProvider: NetworkInfoProvider,
  logger = dummyLogger
): ChainHistoryProvider => {
  const currentWalletProtocolParameters = await networkInfoProvider.currentWalletProtocolParameters();

  const transactionsByHashes: ChainHistoryProvider['transactionsByHashes'] = async (hashes) => {
    return Promise.all(hashes.map((hash) => fetchTransaction(hash, currentWalletProtocolParameters)));
  }

  return { transactionsByHashes };
}
```

@mkazlauskas thoughts?

Reply:
Agree, providers can depend on other providers :+1:

Reply:
For now I added a `TODO` comment to move `currentWalletProtocolParameters` to `NetworkInfoProvider`

Reply:
ADP-1856